lib/swagger/swagger.ex

```diff
@@ -2,127 +2,122 @@ defmodule Xcribe.Swagger do
-    swagger_json()
-    |> add_requests(requests)
-    |> add_security(requests)
+    requests
+    |> Enum.sort_by(& &1.status_code)
+    |> build_openapi_object()
+    |> include_security_scheme(requests)
     |> Xcribe.JSON.encode!()
   end
 
-  defp swagger_json() do
+  defp build_openapi_object(requests) do
     %{
       "openapi" => "3.0.0",
       "info" => %{
         "title" => Map.get(xcribe_info(), :name, ""),
         "version" => Map.get(xcribe_info(), :version, "0.1.0"),
         "description" => Map.get(xcribe_info(), :description, "")
-      }
+      },
+      "paths" => paths_from_requests(requests)
     }
   end
 
-  defp add_requests(swagger_map, requests) do
-    paths =
-      requests
-      |> Enum.sort_by(& &1.status_code)
-      |> Enum.reduce(%{}, fn x, acc ->
-        Map.put(acc, x.path, Map.merge(acc[x.path] || %{}, handle_request(x, acc)))
-      end)
+  defp paths_from_requests(requests),
+    do: Enum.reduce(requests, %{}, &include_request_into_paths/2)
 
-    Map.put(swagger_map, "paths", paths)
+  defp include_request_into_paths(%{path: path} = request, paths) do
+    Map.update(
+      paths,
+      path,
+      build_path_item_object(request),
+      &include_request_into_given_path(&1, request)
+    )
   end
 
-  defp add_security(swagger_map, requests) do
-    requests
-    |> Enum.any?(&has_authorization_header?/1)
-    |> if do
-      Map.put(swagger_map, "components", %{
-        "securitySchemes" => %{
-          "api_key" => %{"name" => "Authorization", "type" => "apiKey", "in" => "header"}
-        }
-      })
-    else
-      swagger_map
-    end
+  defp include_request_into_given_path(path, %{verb: verb} = request) do
+    Map.update(
+      path,
+      verb,
+      build_operation_object(request),
+      &include_request_into_given_verb(&1, request)
+    )
   end
 
-  defp has_authorization_header?(request) do
-    request
-    |> Map.fetch!(:header_params)
-    |> Enum.any?(fn {header, _} -> String.downcase(header) == "authorization" end)
+  defp include_request_into_given_verb(operation_object, request) do
+    Map.update(operation_object, "responses", %{}, fn responses ->
+      Map.merge(responses, Formatter.format_responses(request))
+    end)
   end
 
-  defp handle_request(request, swagger_paths) do
-    swagger_paths
-    |> Map.fetch(request.path)
-    |> has_key?(request.verb)
-    |> handle_request(swagger_paths, request)
-  end
-
-  defp has_key?(:error, _), do: false
-  defp has_key?({:ok, map}, key), do: Map.has_key?(map, key)
-
-  defp handle_request(true, swagger_paths, request), do: add_response(swagger_paths, request)
-  defp handle_request(false, _swagger_paths, request), do: format_request(request)
-
-  defp add_response(swagger_map, request) do
-    original_request = swagger_map[request.path][request.verb]
+  defp build_path_item_object(%{verb: verb} = request),
+    do: %{verb => build_operation_object(request)}
 
+  defp build_operation_object(request) do
     %{
```

Comment:
you can extract this map to a function to make all function in this pipe in one layer of abstraction

```elixir
request
|> prepare_open_api_request
|> parameters_if_needed(request)
|> request_body_if_needed(request)
|> security_if_needed(request)

# ...
def prepare_open_api_request(request) do
   %{
     "summary" => Descriptor.get_action_description(request),
      "description" => Descriptor.get_request_description(request),
      "responses" => Formatter.format_responses(request)
    }
end
```

Reply:
Done :+1: